### PR TITLE
Skeleton spawner

### PR DESCRIPTION
**Pull request by Dtinpa**

### CODE REVIEW

File: Assets/Code/Runtime/SkeletonSpawner.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    [SerializeField]
+    private float spawnCollisionDetectionRadius = 1f;
+    [SerializeField]
+    private List<string> excludeSpawnCollisionTags;
+    [SerializeField]
+    private int spawnRateIncreaseByFactor = 2;
+ 
+    private int skeletonCount = 0;
+
+    //if the global spawn limit has been reached, stops spawning enemies entirely
```
Comment (brandoncimino) [inline Assets/Code/Runtime/SkeletonSpawner.cs]:
```suggestion
    /// if the global spawn limit has been reached, stops spawning enemies entirely
```

File: Assets/Code/Runtime/SkeletonSpawner.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    private bool continueSpawning = true;
+    //if we tried to spawn something but it was still colliding, then skip instantiating so the game doesn't freeze
```
Comment (brandoncimino) [inline Assets/Code/Runtime/SkeletonSpawner.cs]:
```suggestion
    /// if we tried to spawn something but it was still colliding, then skip instantiating so the game doesn't freeze
```

File: Assets/Code/Runtime/SkeletonSpawner.cs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    private bool couldSpawn = true;
+    //if this spawner is the closest to the player, then speed up the spawn rate
```
Comment (brandoncimino) [inline Assets/Code/Runtime/SkeletonSpawner.cs]:
```suggestion
    /// if this spawner is the closest to the player, then speed up the spawn rate
```

File: Assets/Code/Runtime/SkeletonSpawner.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            } else
+            {
+                couldSpawn = true;
+            }
+        }
+
+        StartCoroutine(spawnEnemy(enemy, interval, enemyMaxSpawn));
+    }
+
+    private Vector3 determineSpawnLocation()
```
Comment (brandoncimino) [inline Assets/Code/Runtime/SkeletonSpawner.cs]:
```suggestion
        private static RaycastHit HitTerrain(float x, float z) {
        var rayOrigin = new Vector3(x, 0, z);
        return Physics.Raycast(rayOrigin, Vector3.down, out var hit) ? hit : throw new InvalidOperationException($"Point (x: {x}, z: {z}) isn't in this terrestrial sphere!");
    }

    private Vector3? determineSpawnLocation() {
        var spawnX   = Random.Range(xAxisMaxRange, xAxisMaxRange);
        var spawnZ   = Random.Range(zAxisMaxRange, zAxisMaxRange);
        var position = transform.position;
        var spawnPos = HitTerrain(position.x + spawnX, position.z + spawnZ).point;

        const int maxLoops = 5;
        for (int i = 0; i < maxLoops; i++) {
            if (spawnIsColliding(spawnPos) == false) {
                return spawnPos;
            }
            
            spawnPos.x = (this.transform.position.x > spawnPos.x) ? spawnPos.x + xCollisionOffset : spawnPos.x - xCollisionOffset;
            spawnPos.z = (this.transform.position.z > spawnPos.z) ? spawnPos.z + zCollisionOffset : spawnPos.z - zCollisionOffset;
        }

        return null;
    }
```

**Comment by Dtinpa** [comment]
Does this look good to merge in?

**Review by brandoncimino — APPROVED** [review]

**Pull request merged**